### PR TITLE
NIFI-4833 Add ScanHBase processor

### PR DESCRIPTION
**Pull request by bdesert**
### Description:

Add new processor ScanHBase and a test package.

------------------------------------------------------
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by MikeThomsen** [comment]
I'll take a stab at reviewing, but you should git cherry pick this onto a git branch and not do it off your master branch.

**Comment by MikeThomsen** [comment]
I have a patch that I've been working on for adding support for HBase visibility labels to the existing processors. Might want to think about how to integrate that into this processor.

File: nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+    
+    static final PropertyDescriptor REVERSED_SCAN = new PropertyDescriptor.Builder()
+            .displayName("Reversed order")
+            .name("scanhbase-reversed-order")
+            .description("Set whether this scan is a reversed one. This is false by default which means forward(normal) scan.")
+            .expressionLanguageSupported(false)
+            .allowableValues("true", "false")
+            .required(false)
+            .defaultValue("false")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
[the remark this answers is not in the document]
Comment (bdesert) [reply nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java]: Agree. Although it doesn't make any difference (due to allowable values), but looks nicer with boolean validator.

File: nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+            .build();
+    
+    static final PropertyDescriptor FILTER_EXPRESSION = new PropertyDescriptor.Builder()
+            .displayName("Filter expression")
+            .name("scanhbase-filter-expression")
+            .description("An HBase filter expression that will be applied to the scan. This property can not be used when also using the Columns property.")
```
[the remark this answers is not in the document]
Comment (bdesert) [reply nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java]: It's basically regular syntax of hbase shell filters. I don't parse them and just pass to HBase client as is. so if there is any change in future version to the filter i'll provide as an example, that example won't be valid, while processor itself will still be OK. That was my concern. If you still think it's would be better to have an example, I'll add it. Let me know.

File: nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+            .addValidator(StandardValidators.INTEGER_VALIDATOR)
+            .build();
+    
+    
+    static final PropertyDescriptor REVERSED_SCAN = new PropertyDescriptor.Builder()
+            .displayName("Reversed order")
+            .name("scanhbase-reversed-order")
+            .description("Set whether this scan is a reversed one. This is false by default which means forward(normal) scan.")
+            .expressionLanguageSupported(false)
+            .allowableValues("true", "false")
```
[the remark this answers is not in the document]
Comment (bdesert) [reply nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java]: if there were constant defined for boolean types, I would agree. But in this case, I would keep it this way, it's just cleaner and less lines

File: nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+        final List<ValidationResult> problems = new ArrayList<>();
+
+        final String columns = validationContext.getProperty(COLUMNS).getValue();
+        final String filter = validationContext.getProperty(FILTER_EXPRESSION).getValue();
+
+        if (!StringUtils.isBlank(columns) && !StringUtils.isBlank(filter)) {
+            problems.add(new ValidationResult.Builder()
+                    .subject(FILTER_EXPRESSION.getDisplayName())
+                    .input(filter).valid(false)
+                    .explanation("a filter expression can not be used in conjunction with the Columns property")
```
[the remark this answers is not in the document]
Comment (bdesert) [reply nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/ScanHBase.java]: 👍

**Pull request closed without merging**